### PR TITLE
RavenDB-13098 TimeSeries replication - conflict resolution

### PR DESCRIPTION
**Pull request by karmeli87**
silently merge on conflicts with attachment/counters/time-series

* refactor recreation of counters/attachments and also time-series

### CODE REVIEW

File: src/Raven.Server/Documents/AttachmentsStorage.cs [hunk 1]
```diff
@@ -544,6 +544,15 @@ public DynamicJsonArray GetAttachmentsMetadataForDocument(DocumentsOperationCont
             return attachments;
         }
 
+        public (IEnumerable<object>, int) GetAttachmentsMetadataForDocument(DocumentsOperationContext context, string docId)
```
Comment (ayende) [inline src/Raven.Server/Documents/AttachmentsStorage.cs]:
```suggestion
        public (DynamicJsonArray , int) GetAttachmentsMetadataForDocument(DocumentsOperationContext context, string docId)
```

File: src/Raven.Server/Documents/CountersStorage.cs [hunk 2]
```diff
@@ -1200,6 +1201,12 @@ public IEnumerable<string> GetCountersForDocument(DocumentsOperationContext cont
             }
         }
 
+        public (IEnumerable<object> Items, int Count) GetCountersForDocumentList(DocumentsOperationContext context, string docId)
```
Comment (ayende) [inline src/Raven.Server/Documents/CountersStorage.cs]:
```suggestion
        public (DynamicJsonArray  Items, int Count) GetCountersForDocumentList(DocumentsOperationContext context, string docId)
```

File: src/Raven.Server/Documents/DocumentPutAction.cs [hunk 3]
```diff
@@ -23,11 +24,18 @@ public unsafe class DocumentPutAction
...
         private readonly DocumentDatabase _documentDatabase;
+        private readonly IRecreationType[] _recreationTypes;
 
         public DocumentPutAction(DocumentsStorage documentsStorage, DocumentDatabase documentDatabase)
         {
             _documentsStorage = documentsStorage;
             _documentDatabase = documentDatabase;
+            _recreationTypes = new IRecreationType[]
+            {
+                new RecreateAttachment(documentsStorage),
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentPutAction.cs]:
```suggestion
                new RecreateAttachments(documentsStorage),
```

File: src/Raven.Server/Documents/DocumentPutAction.cs [hunk 4]
```diff
@@ -390,164 +389,183 @@ private static void ThrowInvalidDocumentId(string id)
...
-            if (metadata == null)
+            DocumentFlags HasFlag { get; }
+
+            NonPersistentDocumentFlags ResolveConflictFlag { get; }
+            NonPersistentDocumentFlags ByUpdateFlag { get; }
+
+            Action<BlittableJsonReaderObject, DocumentFlags> Assert { get; }
+        }
+
+        private class RecreateAttachment : IRecreationType
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentPutAction.cs]:
```suggestion
        private class RecreateAttachments : IRecreationType
```

File: src/Voron/SliceOptions.cs [hunk 5]
```diff
@@ -7,5 +7,6 @@ public enum SliceOptions : byte
         Key = 0,
         BeforeAllKeys = ByteStringType.UserDefined1,
         AfterAllKeys = ByteStringType.UserDefined2,
+        OutOfScope = ByteStringType.UserDefined3 // this used so seeking on a tree will return false
```
Comment (ayende) [inline src/Voron/SliceOptions.cs]: I don't like this, why do we need this? Can't you use `AfterAllKeys`, if you want this search?

File: src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs [hunk 6]
```diff
@@ -1346,6 +1346,37 @@ private Table GetTimeSeriesTable(Transaction tx, CollectionName collection)
             return tx.OpenTable(TimeSeriesSchema, tableName);
         }
 
+
+      
+
+        public (IEnumerable<object>, int Count) GetTimeSeriesNamesForDocument(DocumentsOperationContext context, string docId)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs]: Return `List<string>` here? You do not need to return tuple.

File: src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs [hunk 7]
```diff
@@ -1346,6 +1346,37 @@ private Table GetTimeSeriesTable(Transaction tx, CollectionName collection)
...
+        {
+            var table = new Table(TimeSeriesSchema, context.Transaction.InnerTransaction);
+            var list = new List<string>();
+            using (DocumentIdWorker.GetSliceFromId(context, docId, out Slice documentKeyPrefix, SpecialChars.RecordSeparator))
+            {
+                var valueSlice = Slices.OutOfScope; // seek for this value will return false, so we will get the last value in the tree
```
Comment (ppekrol) [inline src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs]: Can't we use AfterAllKeys?

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Server/Documents/DocumentPutAction.cs [hunk 8]
```diff
@@ -390,164 +389,182 @@ private static void ThrowInvalidDocumentId(string id)
...
-                        return true;
-                    }
-                }
+                _storage = storage;
             }
 
-            return false;
+            public string MetadataProperty => Constants.Documents.Metadata.TimeSeries;
+
+            public Func<DocumentsOperationContext, string, DynamicJsonArray> GetMetadata => _storage.TimeSeriesStorage.GetTimeSeriesNamesForDocument;
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentPutAction.cs]: Is there a reason this is a property to a `Func` and not a simple method?

File: src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs [hunk 9]
```diff
@@ -1346,6 +1346,48 @@ private Table GetTimeSeriesTable(Transaction tx, CollectionName collection)
...
+            var holder = new ByteStringStorage
+            {
+                Flags = ByteStringType.Mutable,
+                Length = 0,
+                Ptr = (byte*)0,
+                Size = 0
+            };
+            var slice = new Slice(new ByteString(&holder));
+
+            using (DocumentIdWorker.GetSliceFromId(context, docId, out Slice documentKeyPrefix, SpecialChars.RecordSeparator))
```
Comment (ayende) [inline src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs]: I don't follow what you are doing here.

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs [hunk 10]
```diff
@@ -1435,7 +1435,7 @@ public DynamicJsonArray GetTimeSeriesNamesForDocument(DocumentsOperationContext
                 while (table.SeekOneBackwardByPrimaryKeyPrefix(documentKeyPrefix, slice, out reader))
                 {
                     name = UpdateSlice(ref reader, documentKeyPrefix.Size, ref slice);
-                    list.Add(name);
+                    list.Insert(0, name);
```
Comment (ayende) [inline src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs]: That is going to be costly. Better to add, then reverse in the end

File: src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs (hunk 9 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs]: we need to find all of the time series names for a given document.
for example we have the following segments:
```
            doc/heartbeats/123
            doc/heartbeats/234
            doc/heartbeats/666
            doc/pulse/123
            doc/pulse/54656
```
so we seek backwards, starting from the end.
extracting the last name and use it as a prefix for the next iteration

File: src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs (hunk 10 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs]: done

**Review by ppekrol — APPROVED** [review]

**Review by ayende — APPROVED** [review]

**Pull request merged**